tripy/tripy/backend/compiler_api.py

```diff
@@ -31,6 +31,169 @@
+        linear = tp.Linear(2, 3)
+        compiler = tp.Compiler(linear)
+
+        compiled_linear = compiler.compile(tp.InputInfo((2, 2), dtype=tp.float32))
+
+        a = tp.ones((2, 2), dtype=tp.float32)
+
+        compiled_linear.stream = default_stream
+        out = compiled_linear(a)
+        default_stream.synchronize()
+        assert cp.array_equal(cp.from_dlpack(out), cp.from_dlpack(linear(a)))
+
+    Usage:
+        default = default_stream()  # Returns the default Stream object for the current device
+        default_gpu1 = default_stream(tp.device("gpu", 1))  # Returns the default Stream object for GPU 1
+
+    Note:
+        Calling default_stream() with the same device always returns the same Stream instance for that device.
+    """
+    global _default_stream_instances
+
+    if device.kind != "gpu":
+        raise ValueError(f"default_stream creation requires device to be of type gpu, got device={device}.")
+
+    if device.index != 0:
+        raise ValueError(f"Tripy stream only works with device index 0, got device={device}")
+
+    if device.index not in _default_stream_instances:
+        _default_stream_instances[device.index] = Stream()
+
+    return _default_stream_instances[device.index]
+
+
+@export.public_api(document_under="compiler")
+def synchronize() -> None:
+    """
+    Synchronizes the default stream.
+
+    .. code-block:: python
+        :linenos:
+        :caption: Use streams in execution
+
+        linear = tp.Linear(2, 3)
+        compiler = tp.Compiler(linear)
+
+        compiled_linear = compiler.compile(tp.InputInfo((2, 2), dtype=tp.float32))
+
+        a = tp.ones((2, 2), dtype=tp.float32)
+
+        out = compiled_linear(a)
+        tp.synchronize()
+        assert cp.array_equal(cp.from_dlpack(out), cp.from_dlpack(linear(a)))
+    """
+    default_stream().synchronize()
+
+
+@export.public_api(document_under="compiler")
+class Stream:
+    """
+    Represents a CUDA stream that can be used to manage concurrent operations.
+
+    This class is a wrapper around the underlying stream object, allowing management of CUDA streams.
+
+    Args:
+        priority (int): Assign priority for the new stream. Lower number signifies higher priority.
+
+    .. code-block:: python
+        :linenos:
+        :caption: Creating new streams
+
+        streamA = tp.Stream()
+        streamB = tp.Stream()
+        defaultStream = tp.default_stream()
+        assert streamA != streamB
+        assert defaultStream != streamB
+
+    .. code-block:: python
+        :linenos:
+        :caption: Example
+
+        linear = tp.Linear(2, 3)
+        compiler = tp.Compiler(linear)
+
+        compiled_linear = compiler.compile(tp.InputInfo((2, 2), dtype=tp.float32))
+
+        a = tp.ones((2, 2), dtype=tp.float32)
+
+        compiled_linear.stream = tp.Stream()
+        out = compiled_linear(a)
+        assert cp.array_equal(cp.from_dlpack(out), cp.from_dlpack(linear(a)))
+    """
```

Comment:
This part of the docstring should be under the constructor to make it render correctly. Also, we can drop the type annotations